File: leetcode/207_Course_Schedule.py
from typing import List
from unittest import TestCase

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        def rec(v: int, adj_list: List[List[int]], in_stack: List[bool], visited: List[bool]):
            visited[v] = True
            in_stack[v] = True
            for u in adj_list[v]:
                if in_stack[u]:
                    return False
                if not visited[u]:
                    if not rec(u, adj_list, in_stack, visited):
                        return False
            in_stack[v] = False
            return True

        n = numCourses
        adj_list: List[List[int]] = [[] for i in range(n)]
        for v, u in prerequisites:
            adj_list[u].append(v)
        visited, in_stack = [False] * n, [False] * n
        for v in range(n):
            if not visited[v]:
                if not rec(v, adj_list, in_stack, visited):
                    return False
        return True

class CanFinishTest(TestCase):    
    def test_1(self):
        self.assertTrue(Solution().canFinish(2, [[1,0]]))

    def test_2(self):
        self.assertFalse(Solution().canFinish(2, [[1,0],[0,1]]))

File: leetcode/207_Course_Schedule.rs
struct Solution {}

#[derive(Clone, PartialEq)]
enum State {
    NotVisited,
    InStack,
    Done,
}

impl Solution {
    pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        let n = num_courses as usize;
        let mut adj_list = vec![vec![]; n];
        for (v, u) in prerequisites.into_iter().map(|edge| (edge[0] as usize, edge[1] as usize)){
            adj_list[u].push(v);
        }
        let mut stack: Vec<usize>= Vec::with_capacity(n);
        let mut states = vec![State::NotVisited; n];
        for v in 0..n {
            if states[v] == State::NotVisited {
                stack.push(v);
                while let Some(v) = stack.pop() {
                    match states[v] {
                        State::NotVisited => {
                            states[v] = State::InStack;
                            stack.push(v);
                            for &u in adj_list[v].iter() {
                                match states[u] {
                                    State::InStack => { return false; },
                                    State::NotVisited => { stack.push(u); },
                                    State::Done => {}
                                }
                            }
                        },
                        State::InStack => { states[v] = State::Done; },
                        State::Done => {},
                    };
                }
            }            
        }
        true
    }
}

#[test]
fn test_1() {
    assert!(Solution::can_finish(2, vec![vec![1,0]]));
}

#[test]
fn test_2() {
    assert!(! Solution::can_finish(2, vec![vec![1,0],vec![0,1]]));
}
